test/test_common/utility.h

```diff
@@ -620,75 +621,138 @@ class ConditionalInitializer {
 namespace Http {
 
 /**
- * A test version of HeaderMapImpl that adds some niceties around letting us use
- * std::string instead of always doing LowerCaseString() by hand.
+ * All of the inline header functions that just pass through to the child header map.
  */
-class TestHeaderMapImpl : public HeaderMapImpl {
+#define DEFINE_TEST_INLINE_HEADER_FUNCS(name)                                                      \
+public:                                                                                            \
+  const HeaderEntry* name() const override { return header_map_->name(); }                         \
+  void append##name(absl::string_view data, absl::string_view delimiter) override {                \
+    header_map_->append##name(data, delimiter);                                                    \
+  }                                                                                                \
+  void setReference##name(absl::string_view value) override {                                      \
+    header_map_->setReference##name(value);                                                        \
+  }                                                                                                \
+  void set##name(absl::string_view value) override { header_map_->set##name(value); }              \
+  void set##name(uint64_t value) override { header_map_->set##name(value); }                       \
+  void remove##name() override { header_map_->remove##name(); }
+
+/**
+ * Base class for all test header map types. This class wraps an underlying real header map
+ * implementation, passes through all calls, and adds some niceties for testing that we don't
+ * want in the production implementation for performance reasons. The wrapping functionality is
+ * primarily here to deal with complexities around virtual calls in some constructor paths in
+ * HeaderMapImpl.
```

Comment:
Sweet! This means we have legit reasons to wrap all the methods, so we can move `verifyByteSize()` in to the wrappers and remove that prod code. Will do when this PR is merged

Reply:
Yes exactly. I was going to do this in a follow up but happy to defer to you if you want to do it. We can chat after this lands.